File: log_generators/log_gen.rb
# log_gen.rb — Auto-push logger (Ruby)
# Usage:
#   require_relative '../log_generators/log_gen'
#   LogGen.start
#   LogGen.log("kuch bhi")
#   LogGen.stop

require 'digest'
require 'fileutils'
require 'time'

module LogGen
  WS       = ENV['GITHUB_WORKSPACE'] || '.'
  OUT      = "#{WS}/output"
  RUNTIME  = "#{OUT}/runtimelogs.txt"
  DETAILED = "#{OUT}/detailedlog.txt"

  @last_hash = nil
  @thread    = nil

  def self._ts = Time.now.utc.strftime('%Y-%m-%d %H:%M:%S UTC')

  def self._hash
    c = [RUNTIME, DETAILED].map { |f| File.read(f) rescue '' }.join
    Digest::MD5.hexdigest(c)
  end

  def self._git(*args) = system('git', *args, chdir: WS, out: '/dev/null', err: '/dev/null')

  def self._push
    h = _hash
    return if h == @last_hash          # kuch badla nahi — skip!
    @last_hash = h
    _git('add', 'output/')
    return if system('git diff --staged --quiet', chdir: WS, out: '/dev/null')
    _git('commit', '-m', "🔄 #{Time.now.utc.strftime('%H:%M:%S')}")
    _git('pull', '--rebase', 'origin', 'main')
    _git('push')
  end

  def self.log(msg, level = 'INFO')
    FileUtils.mkdir_p(OUT)
    line = "[#{_ts}] [#{level}] #{msg}"
    puts line
    File.open(RUNTIME, 'a') { |f| f.puts line }
  end

  def self.start
    FileUtils.mkdir_p(OUT)
    log('🚀 LogGen started!')
    @thread = Thread.new { loop { sleep 5; _push } }
  end

  def self.stop
    @thread&.kill
    log('🏁 LogGen stopped — final push...')
    _push
  end
end

File: log_generators/log_gen.py
"""
log_gen.py — Auto-push logger (Python)
Usage:
    from log_generators.log_gen import start, log, stop
    start()
    log("kuch bhi")
    stop()
"""
import os, subprocess, threading, hashlib
from datetime import datetime, timezone

_WS       = os.environ.get("GITHUB_WORKSPACE", ".")
_OUT      = f"{_WS}/output"
_RUNTIME  = f"{_OUT}/runtimelogs.txt"
_DETAILED = f"{_OUT}/detailedlog.txt"
_stop     = threading.Event()
_lock     = threading.Lock()
_last_h   = None

# ── helpers ────────────────────────────────────────────────────────────
def _ts():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")

def _hash():
    c = ""
    for f in [_RUNTIME, _DETAILED]:
        try: c += open(f).read()
        except: pass
    return hashlib.md5(c.encode()).hexdigest()

def _git(*args):
    return subprocess.run(["git"] + list(args), cwd=_WS,
                          capture_output=True)

def _push():
    global _last_h
    with _lock:
        h = _hash()
        if h == _last_h: return          # kuch badla nahi — skip!
        _last_h = h
        _git("add", "output/")
        if _git("diff", "--staged", "--quiet").returncode == 0: return
        ts = datetime.now(timezone.utc).strftime("%H:%M:%S")
        _git("commit", "-m", f"🔄 {ts}")
        _git("pull", "--rebase", "origin", "main")
        _git("push")

def _loop():
    while not _stop.wait(5):
        _push()

# ── public API ─────────────────────────────────────────────────────────
def log(msg, level="INFO"):
    os.makedirs(_OUT, exist_ok=True)
    line = f"[{_ts()}] [{level}] {msg}"
    print(line, flush=True)
    with open(_RUNTIME, "a") as f:
        f.write(line + "\n")

def start():
    os.makedirs(_OUT, exist_ok=True)
    log("🚀 LogGen started!")
    threading.Thread(target=_loop, daemon=True).start()

def stop():
    _stop.set()
    log("🏁 LogGen stopped — final push...")
    _push()
